### PR TITLE
C pointer slices are no longer allowzero

### LINKED ISSUE #4401: Slicing `[*c]T` returns `[]allowzero T`
```zig
pub fn main() void {
    var buf: [10]u32 = undefined;
    const clients = @ptrCast([*c]u32, &buf);
    @compileLog(@TypeOf(clients[0..1])); // []allowzero u32

    for (clients[0..1]) |*cl| {
        @compileLog(@TypeOf(cl)); // *allowzero u32
    }

    for (clients[0..1]) |cl| {
        @compileLog(@TypeOf(cl)); // compile error trying to dereference *allowzero
    }
}
```

### PR DESCRIPTION
**Pull request by alexnask**
Instead, a non-null check is added at runtime if the C pointer is allowzero.
Closes #4401

### CODE REVIEW

File: src/ir.cpp [hunk 1]
```diff
@@ -26344,6 +26357,11 @@ static IrInstGen *ir_analyze_instruction_slice(IrAnalyze *ira, IrInstSrcSlice *i
             elem_type = array_type->data.pointer.child_type;
             if (array_type->data.pointer.ptr_len == PtrLenC) {
                 array_type = adjust_ptr_len(ira->codegen, array_type, PtrLenUnknown);
+
+                if (array_type->data.pointer.allow_zero) {
+                    array_type = adjust_ptr_allow_zero(ira->codegen, array_type, false);
+                    generate_non_zero_assert = true;
```
Comment (LemonBoy) [inline src/ir.cpp]: `s/zero/null` and please add a brief comment explaining _why_ the `allowzero` is removed and this check is added

File: src/ir.cpp [hunk 2]
```diff
@@ -26618,8 +26636,14 @@ static IrInstGen *ir_analyze_instruction_slice(IrAnalyze *ira, IrInstSrcSlice *i
     if (type_is_invalid(result_loc->value->type) || result_loc->value->type->id == ZigTypeIdUnreachable) {
         return result_loc;
     }
-    return ir_build_slice_gen(ira, &instruction->base.base, return_type,
-        ptr_ptr, casted_start, end, instruction->safety_check_on, result_loc);
+
+    if (generate_non_zero_assert) {
+        IrInstGen *c_ptr_val = ir_get_deref(ira, &instruction->base.base, ptr_ptr, nullptr);
```
Comment (LemonBoy) [inline src/ir.cpp]: `s/c_ptr_val/ptr_val`, this code path is not specific for C pointers (even though atm we only trigger it for C pointers).
Some error checking is also good to have:
```zig
        if (type_is_invalid(ptr_val->value->type))
            return ira->codegen->invalid_inst_gen;
```

File: src/codegen.cpp [hunk 3]
```diff
@@ -979,7 +979,7 @@ static Buf *panic_msg_buf(PanicMsgId msg_id) {
         case PanicMsgIdSliceWidenRemainder:
             return buf_create_from_str("slice widening size mismatch");
         case PanicMsgIdUnwrapOptionalFail:
-            return buf_create_from_str("attempt to unwrap null");
+            return buf_create_from_str("attempt to use null pointer");
```
Comment (Vexu) [inline src/codegen.cpp]: Unwrapping null is not specific to pointers.
```zig
var a: ?u32 = null;
_ = a.?; // 'attempt to use null pointer' ???
```

Comment (alexnask) [reply src/codegen.cpp]: True

**Pull request merged**